File: backend/app/modules/tasks/schemas/ticket.py
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any, Union
from datetime import datetime

from app.modules.tasks.models.ticket import TicketStatus, TicketPriority, TicketType

# 附件可以是字符串（本平台手动上传流程存的是 object_path 字符串），
# 也可以是字典（外部任务源/微信会话写入的 {path, size, filename} 结构）。
# 二者在 tasks/task_comments.attachments(JSON) 列中并存，故统一用联合类型承接。
AttachmentItem = Union[str, Dict[str, Any]]


class TicketBase(BaseModel):
    title: str = Field(..., description="工单标题")
    description: str = Field(..., description="工单描述")
    ticket_type: TicketType = Field(default=TicketType.PROBLEM, description="工单类型")
    priority: TicketPriority = Field(default=TicketPriority.MEDIUM, description="工单优先级")
    customer: Optional[str] = Field(None, description="客户信息")
    team: Optional[str] = Field(None, description="所属团队")
    project_name: Optional[str] = Field(None, description="项目名称")
    project_id: Optional[str] = Field(None, description="项目ID")
    related_resource_id: Optional[int] = Field(None, description="关联资源ID")
    tags: Optional[List[str]] = Field(None, description="标签列表")
    metadata_info: Optional[Dict[str, Any]] = Field(None, description="扩展元数据")
    attachments: Optional[List[AttachmentItem]] = Field(None, description="附件列表")
    deadline_at: Optional[datetime] = Field(None, description="截止时间")


class TicketCreate(TicketBase):
    assigned_to: Optional[str] = Field(None, description="处理者ID")


class TicketUpdate(BaseModel):
    title: Optional[str] = Field(None, description="工单标题")
    description: Optional[str] = Field(None, description="工单描述")
    status: Optional[TicketStatus] = Field(None, description="工单状态")
    priority: Optional[TicketPriority] = Field(None, description="工单优先级")
    ticket_type: Optional[TicketType] = Field(None, description="工单类型")
    assigned_to: Optional[str] = Field(None, description="处理者ID")
    customer: Optional[str] = Field(None, description="客户信息")
    team: Optional[str] = Field(None, description="所属团队")
    project_name: Optional[str] = Field(None, description="项目名称")
    project_id: Optional[str] = Field(None, description="项目ID")
    related_resource_id: Optional[int] = Field(None, description="关联资源ID")
    tags: Optional[List[str]] = Field(None, description="标签列表")
    metadata_info: Optional[Dict[str, Any]] = Field(None, description="扩展元数据")
    attachments: Optional[List[AttachmentItem]] = Field(None, description="附件列表")
    resolved_at: Optional[datetime] = Field(None, description="解决时间")
    deadline_at: Optional[datetime] = Field(None, description="截止时间")
    # 用于操作日志识别，不入库
    operation_type: Optional[str] = Field(None, description="操作类型：escalate/return/reassign/update")


class TicketCommentBase(BaseModel):
    content: str = Field(..., description="评论内容")
    is_public: bool = Field(default=True, description="是否公开")
    attachments: Optional[List[AttachmentItem]] = Field(None, description="附件列表")
    reply_to: Optional[int] = Field(None, description="引用的评论ID（消息引用/回复）")


class TicketCommentCreate(TicketCommentBase):
    pass


class TicketCommentUpdate(BaseModel):
    content: Optional[str] = Field(None, description="评论内容")
    is_public: Optional[bool] = Field(None, description="是否公开")
    attachments: Optional[List[AttachmentItem]] = Field(None, description="附件列表")


class QuotedComment(BaseModel):
    id: int
    content: str
    created_by_name: Optional[str] = None


class TicketCommentResponse(TicketCommentBase):
    id: int
    ticket_id: int
    created_by: str
    created_by_name: Optional[str] = None
    created_by_avatar_resource_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    quoted: Optional[QuotedComment] = Field(None, description="被引用评论的简要信息")

    class Config:
        from_attributes = True


class RedispatchCandidate(BaseModel):
    """R2 重派弹窗候选（精排 Top10 快照）"""
    rank: int = Field(..., description="按权重顺序的排名")
    engineer_id: str = Field(..., description="工程师 users.id")
    name: str = Field(..., description="工程师姓名")
    department: Optional[str] = Field(None, description="部门")
    job_level: Optional[int] = Field(None, description="职级")
    modules: Optional[List[str]] = Field(None, description="责任模块")
    duty: Optional[str] = Field(None, description="职责一句话")
    # 画像缺失英文字段（department/job_level/responsibility_modules），供前端权威判定"待补充画像"
    missing: Optional[List[str]] = Field(None, description="缺失画像字段（全空数组=画像完整）")
    scores: Optional[Dict[str, float]] = Field(None, description="各维度分 {llm,semantic,history,total}")
    tags: Optional[List[str]] = Field(None, description="标记如 项目对接人/上次倾向")


class RedispatchProfile(BaseModel):
    """被派人画像 + 完整性"""
    dept: Optional[str] = None
    job_level: Optional[int] = None
    modules: Optional[List[str]] = None
    duty: Optional[str] = None
    missing: Optional[List[str]] = Field(None, description="缺失画像字段（缺则为空数组）")


class RedispatchResult(BaseModel):
    """R3 派单结果信息（结果卡片/提醒数据源）"""
    assigned_id: str = Field(..., description="实际接单人 users.id")
    assigned_name: Optional[str] = Field(None, description="实际接单人姓名")
    preferred_id: Optional[str] = Field(None, description="意向处理人 users.id（首次派单可为 None）")
    preferred_name: Optional[str] = Field(None, description="意向处理人姓名")
    confidence: Optional[float] = Field(None, description="置信度（拼音命中略降 0.85）")
    decision_type: Optional[str] = Field(None, description="auto/recommend/fallback")
    reasoning: Optional[str] = Field(None, description="派单理由")
    profile: Optional[RedispatchProfile] = Field(None, description="被派人画像+缺失字段（R4 补画像用）")
    matched_pref: Optional[bool] = Field(None, description="是否派到意向人")
    name_collision: Optional[bool] = Field(None, description="是否同名命中（同名提醒）")
    pinyin_match: Optional[bool] = Field(None, description="是否拼音近似名命中（近似名提醒）")
    tip_detail: Optional[str] = Field(None, description="未派到指定人时的完整情商话术（含换人理由与重新派单引导，仅 matched_pref=false 有）")


class TicketRedispatch(BaseModel):
    """详情接口 redispatch 子对象（最新一轮派单完整评估）"""
    dispatch_round: int = Field(..., description="派单轮次")
    candidates: Optional[List[RedispatchCandidate]] = Field(None, description="R2 本轮精排 Top10 快照")
    result: Optional[RedispatchResult] = Field(None, description="R3 本轮派单结果信息")


class TicketResponse(TicketBase):
    id: int
    status: TicketStatus
    redispatch: Optional[TicketRedispatch] = Field(None, description="最新一轮派单评估（无记录为 None）")
    created_by: str
    created_by_name: Optional[str] = None
    assigned_to: Optional[str]
    assigned_to_name: Optional[str] = None
    reporter_name: Optional[str] = None
    assignee_name: Optional[str] = None
    customer: Optional[str]
    customer_name: Optional[str] = None
    team: Optional[str]
    project_name: Optional[str]
    created_at: datetime
    updated_at: datetime
    resolved_at: Optional[datetime]
    closed_at: Optional[datetime]
    deadline_at: Optional[datetime]
    # --- 协商阶段（工单阶段性处理：当前节点 + 节点结束时间）---
    curr_step_id: Optional[int] = Field(None, description="当前协商节点ID")
    curr_step_name: Optional[str] = Field(None, description="当前协商节点名称")
    curr_step_endtime: Optional[datetime] = Field(None, description="当前协商节点结束时间（naive UTC）")
    reply_count: int
    view_count: int
    comments: Optional[List[TicketCommentResponse]] = []

    class Config:
        from_attributes = True


class TicketListItemResponse(TicketBase):
    id: int
    status: TicketStatus
    created_by: str
    created_by_name: Optional[str] = None
    assigned_to: Optional[str]
    assigned_to_name: Optional[str] = None
    customer: Optional[str]
    customer_name: Optional[str] = None
    team: Optional[str]
    project_name: Optional[str]
    created_at: datetime
    updated_at: datetime
    resolved_at: Optional[datetime]
    closed_at: Optional[datetime]
    deadline_at: Optional[datetime]
    curr_step_id: Optional[int] = Field(None, description="当前协商节点ID")
    curr_step_name: Optional[str] = Field(None, description="当前协商节点名称")
    curr_step_endtime: Optional[datetime] = Field(None, description="当前协商节点结束时间（naive UTC）")
    reply_count: int
    view_count: int
    redispatch_tip: Optional[str] = Field(None, description="派单结果提醒一句话摘要（无提醒为 None，见 §3.6）")

    class Config:
        from_attributes = True


class TicketListResponse(BaseModel):
    items: List[TicketListItemResponse]
    total: int
    page: int
    size: int
    pages: int


class TicketQueryParams(BaseModel):
    page: int = Field(default=1, ge=1, description="页码")
    size: int = Field(default=10, ge=1, le=100, description="每页数量")
    id: Optional[int] = Field(None, description="工单ID")
    id_op: Optional[str] = Field(None, description="工单ID过滤操作：equals|gt|gte|lt|lte|ne")
    title: Optional[str] = Field(None, description="工单标题")
    title_op: Optional[str] = Field(None, description="标题过滤操作：equals|contains|notEquals")
    status: Optional[str] = Field(None, description="工单状态，支持多个状态用逗号分隔")
    priority: Optional[TicketPriority] = Field(None, description="工单优先级")
    ticket_type: Optional[TicketType] = Field(None, description="工单类型")
    created_by: Optional[str] = Field(None, description="创建者ID")
    created_by_op: Optional[str] = Field(None, description="创建者过滤操作")
    created_by_name: Optional[str] = Field(None, description="创建者姓名")
    assigned_to: Optional[str] = Field(None, description="处理者ID")
    assigned_to_op: Optional[str] = Field(None, description="处理者过滤操作")
    assigned_to_name: Optional[str] = Field(None, description="处理者姓名")
    customer: Optional[str] = Field(None, description="客户信息")
    customer_op: Optional[str] = Field(None, description="客户过滤操作")
    customer_name: Optional[str] = Field(None, description="客户姓名")
    keyword: Optional[str] = Field(None, description="关键词搜索")
    related_resource_id: Optional[int] = Field(None, description="关联资源ID")
    related_resource_id_op: Optional[str] = Field(None, description="关联资源ID过滤操作")
    tag: Optional[str] = Field(None, description="标签过滤")
    project_name: Optional[str] = Field(None, description="项目名称")
    project_name_op: Optional[str] = Field(None, description="项目名称过滤操作")
    project_id: Optional[str] = Field(None, description="项目ID")
    project_id_op: Optional[str] = Field(None, description="项目ID过滤操作")
    source: Optional[str] = Field(None, description="任务来源：manual/zentao/...")
    source_op: Optional[str] = Field(None, description="来源过滤操作")
    deadline_at: Optional[datetime] = Field(None, description="截止时间")
    created_at_start: Optional[datetime] = Field(None, description="创建时间起始")
    created_at_end: Optional[datetime] = Field(None, description="创建时间结束")
    updated_at_start: Optional[datetime] = Field(None, description="更新时间起始")
    updated_at_end: Optional[datetime] = Field(None, description="更新时间结束")
    resolved_at_start: Optional[datetime] = Field(None, description="解决时间起始")
    resolved_at_end: Optional[datetime] = Field(None, description="解决时间结束")
    closed_at_start: Optional[datetime] = Field(None, description="关闭时间起始")
    closed_at_end: Optional[datetime] = Field(None, description="关闭时间结束")
    deadline_at_start: Optional[datetime] = Field(None, description="截止时间起始")
    deadline_at_end: Optional[datetime] = Field(None, description="截止时间结束")


class TicketCuibanNotification(BaseModel):
    ticket_id: Optional[int] = Field(None, description="工单ID")
    notify_type: int = Field(..., description="通知类型")
    assigned_to: Optional[str] = Field(None, description="处理者ID")
    to_admin: Optional[bool] = Field(False, description="是否通知管理员")


class TicketCreateNotificationRequest(BaseModel):
    """新建工单通知请求（内部接口，供 AI 派单服务调用）。
    调用方只传 task_id，后端查库组装标题/项目/截止时间/受理人等完整字段后发通知。"""
    task_id: int = Field(..., description="工单ID")


class TicketFilter(BaseModel):
    field: Optional[str] = Field(None, description="字段名")
    op: Optional[str] = Field(None, description="运算符")
    value: Optional[Any] = Field(None, description="过滤值")
    or_conditions: Optional[List['TicketFilter']] = Field(None, alias='or', description="OR条件列表")
    and_conditions: Optional[List['TicketFilter']] = Field(None, alias='and', description="AND条件列表")

    class Config:
        populate_by_name = True


class TicketSort(BaseModel):
    field: str = Field(..., description="排序字段")
    direction: str = Field(..., description="排序方向：asc 或 desc")


class TicketFilterRequest(BaseModel):
    filters: List[TicketFilter] = Field(default_factory=list, description="过滤条件列表")
    sorts: Optional[List[TicketSort]] = Field(default_factory=list, description="排序条件列表")
    page: int = Field(default=1, ge=1, description="页码")
    size: int = Field(default=10, ge=1, le=100, description="每页数量")


class ProjectMemberResponse(BaseModel):
    """项目成员（用于 @ 提及选择），复用 user_project_roles 表数据。"""
    id: str
    username: str
    name: Optional[str] = None
    role_name: Optional[str] = None

File: frontend/src/api/ticket.ts
/**
 * 工单 API —— /api/tasks/*
 *
 * - 催办：POST /cuiban-notification（现成，TicketDetailPage handleUrge 已用）
 * - 评论：GET/POST /{ticket_id}/comments（现成，评论绑定 ticket_id）
 * - 附件：POST /comments/attachments（temp_id 关联，MinIO 存储，发评论时一并入库）
 * - 创建：POST /（系统任务创建，兜底双工单场景下工单2 直接指定 assigned_to=项目负责人）
 */
import { createRequest, getToken } from '@/api/client';
import API_CONFIG from '@/config/api';
import { readStored } from '@/stores/authStorage';

const request = createRequest(API_CONFIG.TASKS.BASE_URL, '工单服务');

export interface TicketComment {
  id: number;
  ticket_id: number;
  content: string;
  is_public: boolean;
  /** 附件列表（后端存的是 object_path 字符串数组，如 ["bucket/temp/file.png"]） */
  attachments?: unknown[];
  created_by: string;
  created_by_name?: string;
  created_at: string;
  updated_at: string;
}

/** 工单类型（与后端 TicketType 枚举对齐） */
export type TicketType = 'problem' | 'bug' | 'feature' | 'support' | 'other';
/** 工单优先级（与后端 TicketPriority 枚举对齐） */
export type TicketPriority = 'low' | 'medium' | 'high' | 'urgent';

export interface CreateTicketParams {
  title: string;
  description: string;
  ticket_type?: TicketType;
  priority?: TicketPriority;
  project_name?: string;
  project_id?: string;
  assigned_to?: string;       // 接单人 users.id（过渡期后端也认 username）
  deadline_at?: string;       // 最晚解决时间（ISO 字符串，兜底双工单工单2 复用弹窗选择值）
  customer?: string;
  metadata_info?: Record<string, unknown>;
  tags?: string[];
  /** 附件列表：字符串为 object_path；dict 为 {path, object_path, filename} 结构（远程截图等）。
   *  与 tasks.attachments 列约定对齐——详情页读 path，AI 路径去重读 object_path。 */
  attachments?: Array<string | { path?: string; object_path?: string; filename?: string; [k: string]: unknown }>;
  /** 代他人提单：被代理人 users.id（留空=普通自提单）。
   *  后端会二次校验其为在职用户；成功后建立 pending 代提关系并通知对方。 */
  on_behalf_of?: string;
  /** 初始协商节点 ID（留空=后端按 ticket_type 自动取模板第一个） */
  curr_step_id?: number;
  /** 初始协商节点截止时间（ISO 字符串，留空=后端兜底 deadline_at 或 +7 天） */
  curr_step_endtime?: string;
}

export interface CreatedTicket {
  id: number;
  title: string;
  assigned_to?: string | null;
  [k: string]: unknown;
}

/** 一键催办：通知指定用户 */
export const urgeTicket = (ticketId: number | string, assignedTo: string) =>
  request('/cuiban-notification', {
    method: 'POST',
    body: JSON.stringify({ ticket_id: Number(ticketId), notify_type: 1, assigned_to: assignedTo }),
  });

/** 上报：通知指定用户 + 管理员 */
export const reportTicket = (ticketId: number | string, assignedTo: string) =>
  request('/cuiban-notification', {
    method: 'POST',
    body: JSON.stringify({ ticket_id: Number(ticketId), notify_type: 1, assigned_to: assignedTo, to_admin: true }),
  });

/** 撤回：将工单状态置为已取消（Canceled）
 *  后端 PATCH /{task_id}/status 的 status 参数用 Body(..., embed=True) 声明，
 *  因此请求体必须是 JSON 对象 { status: "canceled" }（不能是裸字符串），否则 FastAPI 报 422 Field required。 */
export const cancelTicket = (ticketId: number | string) =>
  request(`/${Number(ticketId)}/status`, {
    method: 'PATCH',
    body: JSON.stringify({ status: 'canceled' }),
  });

/** 重新派单：强制工单回到待派单状态并触发 AI 智能派单重新推荐处理人。
 *  preferredAssignee 为用户倾向的派单人（users.id，必填）；remark 为可选备注。 */
export const reDispatchTicket = (
  ticketId: number | string,
  preferredAssignee: string,
  remark?: string,
) =>
  request(`/${Number(ticketId)}/re-dispatch`, {
    method: 'POST',
    body: JSON.stringify({ preferred_assignee: preferredAssignee, remark: remark || null }),
  });

// ── 代他人提单（代理提单）：见 docs/PRODUCT/代他人提单（代理提单）功能设计方案.md ──

/** 代理关系状态（与后端 3 态状态机对齐；无接手 / 无 revoke） */
export type ProxyRelationStatus = 'pending' | 'acknowledged' | 'declined';

/** 代提选人候选项：group 由后端显式下发，前端**不靠顺序猜** */
export interface OnBehalfCandidate {
  id: string;
  username: string;
  name?: string | null;
  group: 'project' | 'all';
}

/** 代理关系（详情页横幅 / 列表胶囊用） */
export interface ProxyRelation {
  id: number;
  task_id: number;
  relation_status: ProxyRelationStatus;
  source?: string | null;
  remark?: string | null;
  /** 当前登录用户视角，后端下发，前端不自行拼身份判定 */
  is_agent: boolean;
  is_principal: boolean;
  /** 当前登录用户是否为本单接单人（接单人视角：可见「谁代谁提单」） */
  is_assignee: boolean;
  agent_name?: string | null;
  principal_name?: string | null;
  notified_at?: string | null;
  acked_at?: string | null;
  declined_at?: string | null;
  created_at?: string | null;
}

/** 代提选人候选：同项目在前、其他在职在后（分组标记由后端返回） */
export const getOnBehalfCandidates = (params?: { project_id?: string; keyword?: string }) => {
  const qs = new URLSearchParams();
  if (params?.project_id) qs.set('project_id', params.project_id);
  if (params?.keyword) qs.set('keyword', params.keyword);
  const suffix = qs.toString() ? `?${qs.toString()}` : '';
  return request(`/on-behalf-candidates${suffix}`, { method: 'GET' }) as Promise<OnBehalfCandidate[]>;
};

/** 「待我跟进」角标计数（被代理人、pending 且工单未终结） */
export const getMyFollowupsCount = () =>
  request('/my-followups/count', { method: 'GET' }) as Promise<{ count: number }>;

/** 查询工单的代理关系（非参与人后端返回 403 或空数组） */
export const getProxyRelations = (ticketId: number | string) =>
  request(`/${Number(ticketId)}/proxy-relations`, { method: 'GET' }) as Promise<ProxyRelation[]>;

/** 被代理人确认跟进（pending → acknowledged，获协办权） */
export const ackProxyRelation = (ticketId: number | string, relationId: number) =>
  request(`/${Number(ticketId)}/proxy-relations/${relationId}/ack`, {
    method: 'POST',
    body: JSON.stringify({}),
  }) as Promise<ProxyRelation>;

/** 被代理人拒绝（与我无关，需填原因；工单不中断，代理人兜底推进） */
export const declineProxyRelation = (
  ticketId: number | string,
  relationId: number,
  remark: string,
) =>
  request(`/${Number(ticketId)}/proxy-relations/${relationId}/decline`, {
    method: 'POST',
    body: JSON.stringify({ remark }),
  }) as Promise<ProxyRelation>;

// ── 二次派单感知增强（M2）：详情 redispatch 子对象类型 + 读取 ──
export interface RedispatchCandidate {
  rank: number;
  engineer_id: string;
  name: string;
  department?: string | null;
  job_level?: number | null;
  modules?: string[] | null;
  duty?: string | null;
  // 画像缺失英文字段（department/job_level/responsibility_modules），供前端权威判定"待补充画像"
  missing?: string[] | null;
  scores?: { llm?: number; semantic?: number; history?: number; total?: number } | null;
  tags?: string[] | null;
}

export interface RedispatchProfile {
  dept?: string | null;
  job_level?: number | null;
  modules?: string[] | null;
  duty?: string | null;
  missing?: string[] | null;
  specified_name?: string | null;
}

export interface RedispatchResult {
  assigned_id?: string;
  assigned_name?: string;
  preferred_id?: string | null;
  preferred_name?: string | null;
  confidence?: number | null;
  decision_type?: string | null;
  reasoning?: string | null;
  profile?: RedispatchProfile | null;
  matched_pref?: boolean | null;
  name_collision?: boolean | null;
  pinyin_match?: boolean | null;
  tip_detail?: string | null;
}

export interface TicketRedispatch {
  dispatch_round?: number;
  candidates?: RedispatchCandidate[] | null;
  result?: RedispatchResult | null;
}

/** 读取工单详情的 redispatch 子对象（R2 候选快照 + R3 结果信息），无记录返回 null */
export const fetchRedispatch = (ticketId: number | string) =>
  request<{ redispatch?: TicketRedispatch | null }>(`/${Number(ticketId)}`).then((r) => {
    return r?.redispatch ?? null;
  });

/** 评论列表（按工单绑定） */
export const listComments = (ticketId: number | string) =>
  request<TicketComment[]>(`/${Number(ticketId)}/comments`);

/** 发评论；attachments 为附件上传时用的 temp_id 数组，后端 add_comment 会换成真实路径入库 */
export const addComment = (
  ticketId: number | string,
  content: string,
  attachments: string[] = [],
  isPublic = true,
) =>
  request<TicketComment>(`/${Number(ticketId)}/comments`, {
    method: 'POST',
    body: JSON.stringify({ content, is_public: isPublic, attachments }),
  });

/** 上传评论附件（FormData；temp_id 用于随后发评论时关联）。鉴权带 Bearer token
 *  返回 MinIO 上的真实 object_path（如 "helpdesk/temp/xxx.png"），
 *  前端建单时可直接透传给 createTicket 的 attachments 字段，无需依赖后端进程内存 temp_id 映射。 */
export const uploadCommentAttachment = async (file: File, tempId: string): Promise<string> => {
  const formData = new FormData();
  formData.append('file', file);
  formData.append('temp_id', tempId);
  // 优先取 client.ts 内存 token，其次环境命名空间 localStorage；确保与 createRequest 使用同一个 token
  const token = getToken() || readStored('AUTH_TOKEN') || '';
  const res = await fetch(`${API_CONFIG.TASKS.BASE_URL}/comments/attachments`, {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}` },
    body: formData,
  });
  if (!res.ok) throw new Error(`附件上传失败: ${res.status}`);
  const data = (await res.json()) as { object_path?: string };
  return data.object_path || '';
};

/** 创建系统任务（工单）。返回创建后的工单（含 id）。
 *  用途：兜底双工单场景下，工单2（申请单）走系统任务创建接口，直接指定 assigned_to=项目负责人。
 *  依赖后端改动：create_ticket 尊重 ticket_data.assigned_to（不再硬编码 created_by）。 */
export async function createTicket(params: CreateTicketParams): Promise<CreatedTicket> {
  return request<CreatedTicket>('/', {
    method: 'POST',
    body: JSON.stringify({
      title: params.title,
      description: params.description,
      ticket_type: params.ticket_type ?? 'support',
      priority: params.priority ?? 'medium',
      project_name: params.project_name ?? '',
      project_id: params.project_id ?? '',
      assigned_to: params.assigned_to ?? null,
      deadline_at: params.deadline_at ?? null,
      customer: params.customer ?? null,
      metadata_info: params.metadata_info ?? null,
      tags: params.tags ?? null,
      attachments: params.attachments ?? null,
      curr_step_id: params.curr_step_id ?? null,
      curr_step_endtime: params.curr_step_endtime ?? null,
    }),
  });
}

/** 工单操作日志类型（与后端 OperationType 枚举对齐） */
export type OperationType =
  | 'create'
  | 'status_change'
  | 'assign'
  | 'escalate'
  | 'return'
  | 'reassign'
  | 'update'
  | 'comment'
  | 'view'
  | 'ai_diagnose'
  | 'ai_assign';

export interface OperationLog {
  id: number;
  task_id: number;
  operation_type: OperationType;
  operator: string;
  operator_name?: string | null;
  to_status?: string | null;
  detail?: Record<string, unknown> | null;
  description?: string | null;
  created_at: string;
  ended_at?: string | null;
  duration_seconds?: number | null;
}

/** 获取工单操作日志列表（按时间倒序） */
export const getOperationLogs = (taskId: number | string) =>
  request<OperationLog[]>(`/${Number(taskId)}/operation-logs`, { skipCache: true });

/** 将秒数格式化为人类可读的停留时长，如 "5 分 30 秒" / "45 秒" / "1 小时 5 分" */
export const formatDuration = (seconds: number | null | undefined): string => {
  if (!seconds || seconds <= 0) return '';
  const s = Math.floor(seconds);
  if (s < 60) return `${s} 秒`;
  const h = Math.floor((s / 3600));
  const m = Math.floor((s % 3600) / 60);
  const rest = s % 60;
  if (h > 0) return rest > 0 ? `${h} 小时 ${m} 分` : `${h} 小时 ${m} 分`;
  return rest > 0 ? `${m} 分 ${rest} 秒` : `${m} 分钟`;
};

// ── 工单关联（task_relations）类型与 API ──

/** 关系类型（与后端 RelationType 枚举对齐） */
export type RelationType = 'predecessor' | 'duplicate' | 'subtask';

/** 关联工单概要（target/source 侧） */
export interface RelationBrief {
  id: number;
  title: string;
  status: string;
  created_by_name?: string | null;
  assigned_to_name?: string | null;
}

/** 工单关联响应 */
export interface TaskRelation {
  id: number;
  source_task_id: number;
  target_task_id: number;
  relation_type: RelationType;
  created_by?: string | null;
  created_at: string;
  target?: RelationBrief | null;
  source?: RelationBrief | null;
}

/** 阻塞工单信息（状态变更 422 错误返回） */
export interface BlockedTaskInfo {
  task_id: number;
  title: string;
  status: string;
  reason: 'predecessor' | 'subtask';
}

/** 422 阻塞错误体 */
export interface BlockedErrorDetail {
  code: 'blocked_by_related_tasks';
  message: string;
  blocked: BlockedTaskInfo[];
}

/** 获取工单所有关联（双向） */
export const listRelations = (taskId: number | string) =>
  request<TaskRelation[]>(`/${Number(taskId)}/relations`);

/** 创建工单关联 */
export const createRelation = (
  taskId: number | string,
  targetTaskId: number,
  relationType: RelationType,
) =>
  request<TaskRelation>(`/${Number(taskId)}/relations`, {
    method: 'POST',
    body: JSON.stringify({ target_task_id: targetTaskId, relation_type: relationType }),
  });

/** 删除工单关联 */
export const deleteRelation = (taskId: number | string, relationId: number) =>
  request(`/${Number(taskId)}/relations/${relationId}`, { method: 'DELETE' });

// ── 关系树（树形渲染用） ──

/** 关系树节点 */
export interface RelationTreeNode {
  id: number;
  title: string;
  status: string;
  created_by_name?: string | null;
  assigned_to_name?: string | null;
}

/** 关系树边 */
export interface RelationTreeEdge {
  source: number;
  target: number;
  relation_type: RelationType;
}

/** 关系树响应 */
export interface RelationTreeResponse {
  root_id: number;       // 渲染根节点（subtask 树最顶层父工单）
  current_id: number;     // 用户实际打开的工单
  nodes: RelationTreeNode[];
  edges: RelationTreeEdge[];
}

/** 获取工单完整关系树 */
export const getRelationTree = (taskId: number | string, maxDepth = 8) =>
  request<RelationTreeResponse>(`/${Number(taskId)}/relations/tree?max_depth=${maxDepth}`);
